### ISSUE TITLE
lib file auto-generation script

### ISSUE THREAD

**Issue by dstroy0** [dstroy0/InputHandler#52]
I'm going to start work on one in python, since I don't know much python this'll be fun.  It'll be aimed at user commands.

I'm looking at using tkinter to build the gui, again I don't know how to yet so this may change.

Require the user to copy/paste local/local include the library, add one line to their setup and loop functions. 
It should generate/edit:
InputHandler/src/config/config.h - e
ih_cli/ - g
ih_cli/cli_setup.h - g
ih_cli/cli_commandparameters.h - g
ih_cli/cli_functions.h - g
ih_cli/cli_functions.cpp - g

**Comment by dstroy0** [comment]
I think it should work something like:
user copies library into their sketch directory
user runs this tool and sets their options
tool generates a directory inside of the sketch directory "ih_cli" by default and populates it with files
user adds one line to their setup()
and one line to their loop()

..anything I've missed..

cli just works... and we celebrate?

I think this will be pretty straightforward, I also want to give users the option of saving their options, so probably a json dump to txt in the tool directory.  That'll make it easy for users to share configs.

**Comment by 2bndy5** [comment]
I should point out that python std libs already has a very flexible CLI-based lib named [`argparse`](https://docs.python.org/3/library/argparse.html).

If your aim is to create a python wrapper, then it would help to first make this lib compatible with C++ std libs -- I'm guessing Linux isn't the only platform you're aiming for in a python wrapper.

If you're goal is to create a auto-gen options tool, then that seems like a worthy beginner project. It would involve the `os` lib and the `argparse` std libs. I like python 😉

**Comment by dstroy0** [comment]
Yes, auto-gen the files from the options set in the tool!  I'm using visualtk to start learning how tkinter works, I thought it was a good jumping in point too.

**Comment by 2bndy5** [comment]
CMake also has a mechanism that can be used to generate files based on template files. Most people just use it to inject the current version of a lib into a header (or modify Doxygen's Doxyfile in the same manner).

**Comment by dstroy0** [comment]
I'm thinking that this will fit into two tabs in the app, the first with radio buttons/check boxes/input fields for process setup and the second tab will be for generating CommandParameters command trees, I think that a dropdown tree-list view here would be most appropriate.  Then we can have a menu bar with just file, generate, and about.  File can be open/save/save as, it'll just be a json dump with the script/lib version and options/commands set by the user.  Generate should generate the cli files the directory above InputHandler inside of the user specified directory name or "ih_cli" by default.  About should credit us and link to the git and docs.  Thoughts?

**Comment by 2bndy5** [comment]
It sounds like you want a user project generator. PlatformIO is also written in python and may be useful in the generation process; the [`pio project init`](https://docs.platformio.org/en/latest/core/userguide/project/cmd_init.html) (CLI command) may be very useful. you might be able to use parts of the pio pkg API for a more granular control (like updating existing projects), but I've never needed to use PIO like that.

As for the GUI, it sounds like you have a definitive idea 👍🏼.

**Comment by dstroy0** [comment]
Thanks for the help!

**Comment by dstroy0** [comment]
Python is neat, I don't know nearly enough about environments or how to set them up that slowed me down a lot.  I've had a couple problems trying to use the latest python version with some of the tools.  It seems like I've got my learning environment set up now though.

**Comment by dstroy0** [comment]
Awesome, I've got a couple weeks of learning ahead of me! :)

**Comment by dstroy0** [comment]
![image](https://user-images.githubusercontent.com/20524426/172023378-e324c7b3-2eab-4a21-ba93-5bc71ab8a8d5.png)

The header spacing isn't right for the table, I'll probably do 30/70.  it will look like this on the command tree tab as well but with a tree widget and a Generate button (which is also available in File and as a kb shortcut).  I'm just going to use the same icon as the lib.  I appreciate all your help and input.  I've been playing with the script getting used to different functions and making buttons do things.  Now I'm going through the UI and making object names so I can do that stuff programmatically after designing the ui file and using uic to load it in the script, then hooking the objects and doing the business with them.

**Comment by dstroy0** [comment]
I don't like the whitespace under the code preview, and I think the button group should be centered below the table

**Comment by dstroy0** [comment]
Code preview should be an expanding list, where the list items are file names and it should display the same things no matter which tab you are in

**Comment by 2bndy5** [comment]
Interesting. I've never used QT Design.

**Comment by dstroy0** [comment]
It's making fiddly things like placement super easy.

**Comment by dstroy0** [comment]
So if you've got thoughts or ideas just let me know.  We can add/remove menu items, whatever.  It's easy with this thing, I'm still figuring it out but it's pretty intuitive.

**Comment by 2bndy5** [comment]
My only real GUI experience has been with kivy (python) or flutter (dart).

**Comment by dstroy0** [comment]
Awesome, check out the ui script now.  I haven't hooked into any of the objects yet but it's super easy.  The .ui file is just an xml generated by qt designer

**Comment by dstroy0** [comment]
It's using uic to load right from the ui file

**Comment by 2bndy5** [comment]
I'm not saying you should try anything different (if you want to, then that's fine too). It's probably best to stick with what you're familiar with. Different GUI builders aren't compatible with each other.

Are you working from a local branch? Is there stuff committed that I can dry run? Probably best to keep the UI app in a separate folder at repo root.

I can look into making a CI workflow that will build the UI app... But I need to get my hands dirty first.

**Comment by dstroy0** [comment]
Nope, everything you need is cli_gen_tool; the .pyqt folder at the repo root is generated by an extension for vscode and it just makes objects for me.  The modules you need to run main.py in cli_gen_tool are `PyQt5` and `PyQt5-tools`.

**Comment by dstroy0** [comment]
`PyQt5-tools` gets you Qt designer as well, so you can open the .ui file in designer.  I'm open to learning anything, I just started using this and watched a few videos and it seems easy enough.  You know way more than me; this is my first ui attempt.

**Comment by dstroy0** [comment]
Ok, sorry the `.qt_for_python` is generated by a vscode extension, it uses pyqt uic and pyside6 generates all of the objects based off of the names you pick in qt designer so it saves you some time, you can copy/paste things from there to the main.py in cli_gen_tool to add functionality to buttons etc. If you look in junkyard.py I was manually setting things up in there just to learn about some of the stuff pyqt can do.

**Comment by 2bndy5** [comment]
ok I'm having a peek right now... I have a few questions:
1. Any objection to adding a `venv` folder to the repo's gitignore? This way I can use a dedicated venv for the repo, and VSCode will/should automatically detect and ask to use the venv in the repo.
2. Any objection to restructuring the _cli_gen_tool_ folder into a python package? This can be used for adding _requirements.txt_ and separating out parts of the app into different python scripts (aka "modules").
3. What VSCode ext are you using? I don't see any official QT exts in the VSCode marketplace.
4. I'm still confused about the difference between main.py and cli_tool_gen.py. Is one script for CLI only and the other for GUI only?

**Comment by dstroy0** [comment]
1. Structure away, you know way more about that than me.
2. Same as number one, I'll learn from you how it should be done.
3. https://marketplace.visualstudio.com/items?itemName=seanwu.vscode-qt-for-python `pip install PySide6` this is what generates `main.py` in `.qt_for_python` and THAT `main.py` has all the object hooks for the widgets in `cli_gen_tool/main.py` but it can also launch the ui standalone as well there's just a bunch more modules that way.

**Comment by dstroy0** [comment]
You can get the .qt_for_python dir output by using the compile form option in vscode and selecting the main.ui file as the target.

**Comment by dstroy0** [comment]
4. main.py is generated by qt_for_python from main.ui

cli_gen_tool.py is a uic wrapper that loads main.ui and we can copy some things from main.py into cli_gen_tool.py to save time and learn how to set options correctly. 

Whew I think I got it right that time, this is complicated and I don't understand it fully, sorry.

**Comment by 2bndy5** [comment]
I think using VSCode is better than using QT Creator IDE (not as intuitive as VSCode - for python at least). I've installed the VSCode ext and it generated way more than just _main.py_.

I'm not sure why anything needs to be generated. I'm learning as much as you are when it comes to using QT for a GUI. It looks like PyQTx is a aftermarket lib that consolidates all the PySidex pkg. IDK the purpose of having a "go-between" lib when PySidex would be installed. Is there a inconvenience to implementing both a py file and a ui file?

---
PS - You may have already thought about this, but beginner projects generally don't involve a GUI. However, I'm willing to learn with you...

**Comment by dstroy0** [comment]
I'm just using qt designer to make the ui file, not the qt creator ide, we can use pyqt6 no problem.

I'd like to get the code preview to be a collapsible list of files that will be edited or created based on the options set.

Making the gui seems simple enough...
Generating the code shouldn't be too hard either.

**Comment by 2bndy5** [comment]
I would suggest starting with a CLI only module. Then you can create a GUI that calls functions from the CLI module.

I'll keep reading up on using QT (its very popular). I think I understand now, [this tutorial](https://doc.qt.io/qtforpython/tutorials/basictutorial/uifiles.html) states that the ui file needs to be translated to python code. This approach is a little different from other GUI libs that I've used.

PyQTx isn't advantageous to this app. I think we can simplify things here by not using PyQTx and just stick with PySidex.

I've installed QT v6.3 as that was the latest release. So, I'm using PySide6 for my learning purposes (as that is also suggested in the QT tutorials).

**Comment by 2bndy5** [comment]
Found this [example of a settings editor](https://doc.qt.io/qtforpython/examples/example_corelib__settingseditor.html), but it doesn't use a ui file.

**Comment by dstroy0** [comment]
I wouldn't know where to start with building a cli only version.  That example is really nice, it's the way I would like to display the options and the code preview.  I'm just having fun using the qt designer and learning python!

**Comment by dstroy0** [comment]
> Found this [example of a settings editor](https://doc.qt.io/qtforpython/examples/example_corelib__settingseditor.html), but it doesn't use a ui file.

Pyside generates its code from ui files, then we go in and add functionality to the objects programmatically, is how I understand how the process works so far.

**Comment by dstroy0** [comment]
I added the venv to the gitignore.  Here's the packages I installed to dev the script, if I need PyQt6 I'll grab it too.

```text
(.venv) PS C:\Users\Douglas\Documents\Arduino\libraries\InputHandler> pip list
Package            Version
------------------ -------
pip                22.1.2
PySide6            6.3.0
PySide6-Addons     6.3.0
PySide6-Essentials 6.3.0
setuptools         62.3.3
shiboken6          6.3.0
```

**Comment by dstroy0** [comment]
I pushed a barebones script that loads the ui file so we can reposition things in it or add things and not break the rest of the script

**Comment by 2bndy5** [comment]
> Here's the packages I installed to dev the script

Pretty sure that installing `PySide6` will install all others (even `setuptools` because it's a CPython C-Extension).

I'd recommend that you start a requirements.txt file and put `PySide6` in it. If you find your app using other pip-installable pkgs, then you can add them to that txt file (one pkg per line). This means that all a new dev would have to do is 
```sh
python -m venv env
# activate env
python -m pip install -r cli_gen_tool/requirements.txt
```

**Comment by dstroy0** [comment]
ok, requirements.txt generated from `pip freeze > requirements.txt`

**Comment by dstroy0** [comment]
I also pushed another script, this one is pure python generated from the ui file with qt for python.  We can embed our classes inside of that.  The thing that sucks is that any time the ui file is changed, the mainwindow class gets regenerated so for ui updates our classes would need to be added to the generated mainwindow class again.

**Comment by dstroy0** [comment]
I don't think this will change once it's made though.

**Comment by 2bndy5** [comment]
> ok, requirements.txt generated from `pip freeze > requirements.txt`

Using pip freeze is a bad habit because it includes all dependencies that you shouldn't need to manage. For instance, you pip freeze PySide6 which requires something like `setuptools^=40.0`, but then PySide6 changes their dependencies to require `setuptools>=45`. Managing dependencies of dependencies shouldn't hinder you dev cycle. BTW, the version numbers in my example are wrong, but you get the point.

**Comment by dstroy0** [comment]
Thanks for the warning!  I'm learning a lot, this is fun!

**Comment by dstroy0** [comment]
I connected the button/action objects to methods in `MainWindow` class in `cli_gen_tool.py`.  Now I'm figuring out how to implement the cool settings view you found.

**Comment by dstroy0** [comment]
the about/about and about/help menu items are functional

**Comment by 2bndy5** [comment]
Sounds exciting

**Comment by dstroy0** [comment]
It's pretty cool!  The open file dialog is working.

**Comment by dstroy0** [comment]
I truly have no idea how to structure python, should I break out functions that are longer than 20 or so lines from the class?

**Comment by dstroy0** [comment]
So far everything has been a static function so the methods are short and sweet.

**Comment by 2bndy5** [comment]
20 lines is acceptable in my experience. It gets hard to follow algorithms when the functions are longer than 100 lines or have other functions defined within a function.

There's no explicitly bad practice against defining functions within functions like so:
```py
def func(arg1: int) -> bool:
    """doc string goes here."""

    some_local_var = True

    def local_func() -> bool:
        return some_local_var

    if isinstance(arg1, int):  # proper type checking
        return local_func()
    return False
```

I've really only seen this done when a JS dev tries their hand at python, and it feels just as hacky as it does in JS. Notice I've added type hints in the function definitions; this is a good habit to start doing (but might get complicated if abusing python's dynamic typing). Type hinting does not get used at runtime (its only meant for type checking tools and readabillity).

**Comment by dstroy0** [comment]
I'm very happy you're mentoring me through this.  It feels much less intimidating.  Pretty excited that I got as much done today as I did because I had to read the docs a bunch.

**Comment by dstroy0** [comment]
Is there a better way to get the os name besides https://docs.python.org/3.9/library/os.html#os.uname ?

```python
import os

uname_tuple = os.uname()
os_name = uname_tuple[0] # sysname

```

I haven't ran this yet I've just been thinking about how to make the about hyperlink work cross platform with if elif inside of the about method.

**Comment by 2bndy5** [comment]
`os.uname` is platform specific. Its better to use [`platform.uname()`](https://docs.python.org/3/library/platform.html#platform.uname). It was designed to be a drop-in replacement for better cross-platform code. Here's a test from a python interactive REPL:
```py
>>> import platform
>>> uname_tuple = platform.uname()
>>> print(uname_tuple.system)
Windows
```

**Comment by dstroy0** [comment]
Got a bit done today, it's pretty neat!  I'm figuring out how to use the signals, had to do that to auto update the command length.

**Comment by dstroy0** [comment]
I made an object to put commandParameters into when we pull in all fields in the command settings menu with apply.

**Comment by dstroy0** [comment]
Input validation is very nice, I made it for the function name, command string, and the ids.  the regexp are just temporary until I get the settings_tree menu complete, I'm planning on generating that from config.h after I smoosh config and advanced_config together.  Then parsing that file, and setting those regexp based off of that input.

**Comment by dstroy0** [comment]
We can add items to the tree on InputHandler settings tab, and only the 3rd col is editable by double clicking.  Now I'm going to work on parsing user settings out of config.h.

**Comment by dstroy0** [comment]
Figuring out little bits at a time, this thing is really cool.  I keep getting distracted and going on little tangents.  I'm having so much fun when I'm working on it.  I'm really close to being able to parse the config file the way I want to; but then I learn something new and refactor things.  I absolutely love this part of learning, when everything is new and fresh.  It's like leveling up in a really good rpg or something.  I hope you've been well.  If you've got any thoughts on anything please let me know.  Happy belated father's day if you're a dad.

**Comment by dstroy0** [comment]
parsing the settings out of the config file, they're going to be the 4th match group
`(\s*[\/]*\s*)(\s*#define\s*)(UI_\S*\s*)(\d*)`
and progmem is differentiated like so
`(\s*[\/]*\s*)(\s*#define\s*)(UI_\S*PGM_LEN\s*)(\d*)`

**Comment by dstroy0** [comment]
The settings tree is populating from config.h.  I'm going to put the whole tree into a dict.  Now I have to connect up the edit to the list of strings that makes up config.h, and save everything to the session json.  Pretty cool!

**Comment by 2bndy5** [comment]
Will there also be an option to import a config from json?

**Comment by dstroy0** [comment]
Yes, that's the idea, speed and ease of setup.

**Comment by 2bndy5** [comment]
Awesome! That would definitely help with sharing/tweaking known-good configs. 😃

**Comment by dstroy0** [comment]
I'm adding `cli_gen_tool.json` (the session json) to the ignore list so that a new one is generated on first run.  Also so I don't accidentally push mine out after messing with unimplemented things.

**Comment by 2bndy5** [comment]
Couldn't you just use a virtual json if a json file does not exist? I like the gitignore idea.

**Comment by dstroy0** [comment]
Yes, it's a QFile so I can just
https://doc.qt.io/qt-6/qfile.html#exists-1
```python
file = QFile(path)
        if not file.exists():
            self.session = json.loads(self.defaultGuiOpt)
```

**Comment by dstroy0** [comment]
I really like how I can shorten variable access in python it makes it much more readable

**Comment by dstroy0** [comment]
combo boxes for true/false fields pass their sub dict that contains text fields and the qtreewidgetitem and qcombobox item on index change.  Everything emits a signal when edited in the settings tree now.  Time to work on the code preview for `config.h`.  It will be updated with every item change, which is nice!

**Comment by dstroy0** [comment]
I just did it quick and dirty to get an idea of how to work with the `QPlainTextEdit` object that's doing the preview for us.  It's a one way preview.

**Comment by dstroy0** [comment]
![image](https://user-images.githubusercontent.com/20524426/176280177-6190a00b-ee47-4627-9e60-9d42172cbcba.png)

**Comment by dstroy0** [comment]
I need to add an output category as another root object so that setup.h can be generated

**Comment by 2bndy5** [comment]
This is starting to look close to finished or at least a working proof of concept. I wish the QT implementation for desktops supported the OS's color scheme (light vs dark). It'd also be nice if the text viewer had some syntax highlighting incorporated, but I don't know if that's a task for other libraries like pygments.

**Comment by dstroy0** [comment]
There are rich text widgets and a browser so I think that's possible, the backend is a little messy and needs organized to make it easy but highlighting what the user has changed is possible based on other projects I've looked at, it would just hook into the dict and make a has_changed field true/false or something.  I've been thinking about that and I think that if I can change the copy action to just pick up the plain text that would be best.  Because even though we would like people to press the generate button they might want to copy paste and do their own thing.

**Comment by dstroy0** [comment]
We can change the theme from light to dark as well I don't know if there's a toggle for it programmatically but it's an option I can set in the designer

**Comment by 2bndy5** [comment]
> We can change the theme from light to dark as well I don't know if there's a toggle for it programmatically but it's an option I can set in the designer

If it needs to be hard coded, then I'd vote for dark theme. Its better for those of us spending extraordinary amounts of time looking at screens. Of course, this is your app, so the final decision should be your preference.

**Comment by 2bndy5** [comment]
Everything I use is set to dark theme (when possible).
![image](https://user-images.githubusercontent.com/14963867/176317364-160314a8-b714-43d7-867d-1e6e6bf89a4f.png)

**Comment by dstroy0** [comment]
Yeah all of mine are as well I just hadn't even thought about doing it.

**Comment by dstroy0** [comment]
I can do it with a stylesheet or by setting up a palette manually and passing that to the app in main with `setStyleSheet` or `setPalette` respectively

**Comment by 2bndy5** [comment]
That may be more work than it is worth at this stage. The QT docs have a page with some options: https://wiki.qt.io/Gallery_of_Qt_CSS_Based_Styles

After a quick look, I think
- https://github.com/EClaesson/QTDark might be the easiest solution, but it hasn't been updated in a while. It might be best to use this as a starting point for a CSS.
- https://github.com/ColinDuquesnoy/QDarkStyleSheet seems the most up-to-date, but it requires pyqt and may not be compatible with QT6 (yet).

**Comment by dstroy0** [comment]
I just used someone's module.  The geometry needs adjusting but it looks pretty good.  Dark theme it up.

**Comment by dstroy0** [comment]
There's a ton of little things to do but the basic poc for most things are in place.  I'm going to work on that output section in settings_tree to generate setup.h and then it'll be time to work on generating commands and functions.   Separately, I think everything needs like 5 pixels because of the rounded corners some text is getting cut off.

**Comment by dstroy0** [comment]
the code preview update is updating the wrong text widgets I'll have to look into that tomorrow

**Comment by dstroy0** [comment]
Whatever is happening is what's causing the text to get refreshed in the incorrect `QPlainTextEdit` widget.

**Comment by dstroy0** [comment]
I think it might have something to do with the way I built the dict

**Comment by dstroy0** [comment]
Is this a nono?
```python
generated_filename_subdict = {'filename':'',
                              'file_lines_list':[],
                              'tree_item':{},
                              'contents_item':{},
                              'text_widget':{0:'',1:''}}
#generated_filename_dict = OrderedDict()
generated_filename_dict = {'config.h':generated_filename_subdict,
                           'setup.h':generated_filename_subdict,
                           'parameters.h':generated_filename_subdict,
                           'functions.h':generated_filename_subdict,
                           'functions.cpp':generated_filename_subdict}
```

**Comment by dstroy0** [comment]
Now that's all fixed, I made it so the text pops to where the user made a change like you were talking about, and we can highlight the change individually so that's probably where we can wrap it with some html to change the color.

**Comment by 2bndy5** [comment]
All Python objects are all just pointers under the hood. If you want to use a copy of a dict, then call `<dict_name>.copy()`.

**Comment by dstroy0** [comment]
I tried that, it didn't work.  I think because copy only makes a shallow copy.

**Comment by 2bndy5** [comment]
There is also a `deepcopy()` method in a [std lib called copy](https://docs.python.org/3/library/copy.html). So
```py
from copy import deepcopy

some_dict = dict(key="value")
copied_dict = deepcopy(some_dict)
```

Assignment operators are actually storing the pointer to an object and increasing the object's internal reference counter by 1. The reference counter is how python does garbage collection.

**Comment by dstroy0** [comment]
The method organization is driving me nuts; I keep hesitating organizing them because I merge, cut and paste them constantly but there are so many now it's slowing me down.  I think by reference order in __init__ would probably make the most sense?

**Comment by 2bndy5** [comment]
Yeah I figured the file length was eventually going to be a problem. You could split it up into classes (which could be imported) or just put functions into separate files that are imported as modules.

Using pylint helped me a lot when I was getting the hang of python. Some rules seem ridiculous, but they are born of best practices. There is a way to have vscode use pylint in the python extension settings.

**Comment by 2bndy5** [comment]
BTW, It is pretty common to use [black](https://github.com/psf/black) as a code formatter tool in python. In fact if you right click the python code (anywhere) and select "Format Document", then vscode should automatically ask you if it is ok to install black.

**Comment by dstroy0** [comment]
Cool, I installed the black extension and used it on the code.

**Comment by 2bndy5** [comment]
You'll notice that the indentation is not the same as C++ convention. This is because whitespace is not ignore by the python interpreter (though a C++ compiler does ignore whitespace). So, indentation is rather important, thus it is a part of the [PEP8](https://peps.python.org/pep-0008/) conventional standards.

Using black will let you get on with your life and still adhere to PEP8 standards.

**Comment by 2bndy5** [comment]
I found that using a stub lib for c-extension (like pyside) will help with auto-completion in vscode. Installing https://pypi.org/project/IceSpringPySideStubs-PySide6/ will help with that when using pyside6 API.

A sub lib tells IDEs what data type is used for objects in python code. They also have the benefit of listing the exposing members of modules and data structures.

**Comment by dstroy0** [comment]
Awesome, I'll give that a try too!

**Comment by dstroy0** [comment]
'None' text is being set in column 1 on new rows added.  A tomorrow bug.

**Comment by 2bndy5** [comment]
you can do inline conditionals in python, but they're a little different from C++
```py
some_var = None  # type: str
output = "" if some_var is None else some_var
```

**Comment by 2bndy5** [comment]
dicts can also be expanded as keyword args in function calls:
```py
>>> generated_comment = """/**
...  * @file {file}
...  * @date {date}
...  * @author {author}
...  */"""
>>> default_dict = dict(file="filename.h", date="today", author="me")
>>> output = generated_comment.format(**default_dict)
>>> print(output)
/**
 * @file filename.h
 * @date today
 * @author me
 */
>>>
```

**Comment by dstroy0** [comment]
That's a cool trick!

**Comment by dstroy0** [comment]
Ah, this is what I wanted to do earlier with dicts `dict ={key:**sub_dict,key2:**sub_dict}`; this will reduce setup code a lot.  Thanks!

**Comment by 2bndy5** [comment]
I'm not sure that will work like that. Instead, you should be able to use expansion like so:
```py
sub_dict = dict(key="value", name="description")
main_dict = dict(first=dict(**sub_dict))
# or equivalently
main_dict = {"first": {**sub_dict}}
```
Although I'm not sure it's an alternative to true a deepcopy (nested dicts in the expanded `sub_dict` might use the same pointer under the hood).

**Comment by 2bndy5** [comment]
expansion is really only useful when accepting/passing an arbitrary number of arguments:
```py
def sum(*integers: int) -> int:
    """accepts all args as a list of ints and returns the sum"""
    sum = 0
    for numb in integers:
        sum += numb
    return sum

print(sum(1, 2, 3, 4))
```
the above sample script outputs:
> 10

**Comment by 2bndy5** [comment]
I often expand a tuple to instantiate multiple variables on 1 line
```py
first, second, third = (1, 0b10, b"\x03"[0])
print(first, second, third)
```
which outputs:
> 1, 2, 3

**Comment by dstroy0** [comment]
I'm grouping and helper classing like methods into `/cli_gen_tool/modules/`

I specialize the helper class into the mainwindow class and default init the methods into self.

**Comment by 2bndy5** [comment]
CPython does support multiple inheritence.
```py
class Name:
    def __init__(self):
        self._name = "something"
        super().__init__()

    @property
    def name(self) -> str:
        # instance var cannot have the same name as the property, thus the prefixed '_'
        return self._name

    @name.setter
    def name(self, val: str):
        # setter decorated functions allow us perform extra stuff upon attribute assignments
        self._name = val.strip()  # strip whitespace for example

class Number:
    def __init__(self):
        self.x, self.y = (1, 2)
        super().__init__()

class Base(Name, Number):
    def __init__(self):
        super().__init__()

my_obj = Base()
my_obj.x  # returns 1
my_obj.y  # returns 2
my_obj.name  # returns "something"
my_obj.name = " new thing "
my_obj.name  # returns "new thing"
```

Some use this approach and call it "mixin" classes. Notice the `super().__init__()` calls within the `__init__()` defs; this is important. If passing args to inherited c'tors, then that's a bit more complicated (uses `**kwargs` but alters the dict as keyword args are used from it).

**Comment by dstroy0** [comment]
It's a lot easier to work on and it's not even really organized yet.  I appreciate how easy it was to a) import the external methods and b) add them to the MainWindow instance.

**Comment by 2bndy5** [comment]
When you're ready to start thinking distribution, I can show you how to make it an installable package with an executable script 😉

**Comment by dstroy0** [comment]
drag to resize code preview is working (buggy).  The hitbox is 15x15px in the bottom right corner of each preview pane.  Left click and drag to the new y, release.  The detection is ok, but the redraw needs work, and there needs to be an icon of some kind there to indicate the possibility.

**Comment by dstroy0** [comment]
It's working correctly now, but still needs an indicator.

**Comment by dstroy0** [comment]
I think I need to draw the icon inside of a paintEvent, which I think I can implement in the eventFilter.  I wasn't getting any errors in the test code it just wasn't painting the vertical drag icon in the hit box.  The icon svg is accessible and I can set it as a pixmap icon in columns etc.

**Comment by 2bndy5** [comment]
Instead of allowing users to resize the text boxes couldn't you just make it a list of collapsible items.

<details open><summary>filename.h</summary>

```c
/**
 * file contents
 */
```
</details>
<details><summary>filename.cpp</summary>

```cpp
/**
 * file contents
 */
```
</details>
<details><summary>other_file.h</summary>

```c
/**
 * file contents
 */
```
</details>

**Comment by 2bndy5** [comment]
You can inspect my last comment's raw text by editing it (uses some raw HTML). Typically, Material Design apps sometimes allow only 1 item can be expanded at a time in a collapsible list (when screen real estate is limited).

**Comment by dstroy0** [comment]
That's already happening.  The parent item is collapsible, the container item is now resizable.  I need to paint the icon over the interactive area.

**Comment by dstroy0** [comment]
The scroll area will expand vertically as much as needed but will only display one container widget if it gets enough height.  Then scroll snaps to the next item that's open.  It's pretty slick.

**Comment by 2bndy5** [comment]
oh cool! 🚀  Sorry, I misunderstood what the resizable feature aimed to solve.

**Comment by dstroy0** [comment]
Reading further it looks like I do indeed need to paint that icon inside of the paintEvent for the widget.  There's even a nice warning in the docs that it will only work that way in windows and mac.

**Comment by dstroy0** [comment]
https://srinikom.github.io/pyside-docs/PySide/QtGui/QPainter.html

So, when I was trying to init the QPainter() with the tree object, nothing happened.  Big pink warning explaining why in the docs.

**Comment by dstroy0** [comment]
The cursor changes to vertical resize when hovering over the interactive area

**Comment by dstroy0** [comment]
I finally figured out the syntax to make this work.  I'll try it on the other dicts as well.  So many lines of setup code saved.
```python
## This dict is ordered to preserve insert order for code preview display.
generated_filename_dict = OrderedDict()
## This 'sub' dict is contained by each key in `generated_filename_dict`
generated_filename_sub_dict = {
    "filename": "",
    "file_lines_list": [],
    "tree_item": {},
    "contents_item": {},
    "text_widget": {0: "", 1: ""},    
}
## This dict contains all generated files and associated widgets.
generated_filename_dict = {
    "config.h": {**copy.deepcopy(generated_filename_sub_dict)},
    "setup.h": {**copy.deepcopy(generated_filename_sub_dict)},
    "parameters.h": {**copy.deepcopy(generated_filename_sub_dict)},
    "functions.h": {**copy.deepcopy(generated_filename_sub_dict)},
    "functions.cpp": {**copy.deepcopy(generated_filename_sub_dict)},
}
```

**Comment by 2bndy5** [comment]
You could you just do
```py
## This dict contains all generated files and associated widgets.
generated_filename_dict = {
    "config.h": copy.deepcopy(generated_filename_sub_dict),
    "setup.h": copy.deepcopy(generated_filename_sub_dict),
    "parameters.h": copy.deepcopy(generated_filename_sub_dict),
    "functions.h": copy.deepcopy(generated_filename_sub_dict),
    "functions.cpp": copy.deepcopy(generated_filename_sub_dict),
}
```
and get the same result. `{**some_dict}` is like expanding and packing at the same time. Its resource intensive.
